File: example/customersnode/src/interfaces.ts
import { Expose } from "class-transformer";

export interface LogicalAddress {
  readonly type: string;
  readonly id: string;
  toString(): string;
}

export declare type ClassConstructor<T> = {
  new (...args: any[]): T;
};

export interface Context {
  readonly self: LogicalAddress;
  get<T>(key: string, cls: ClassConstructor<T>): Promise<T | undefined>;
  set<T>(key: string, data: T): void;
  remove(key: string): void;
}

export interface IPublisher<T> {
  subscribe(subscriber?: Partial<ISubscriber<T>>): void;
  map<R>(fn: (data: T) => R): IPublisher<R>;
  forEach(onNextFn: (next: T) => void, requestN?: number): void;
}

export type Source<I> = (subscriber: ISubscriber<I>) => void;

export interface ISubscriber<T> {
  onComplete(): void;
  onError(error: Error): void;
  onNext(value: T): void;
  onSubscribe(subscription: ISubscription): void;
}

export interface ISubscription {
  cancel(): void;
  request(n: number): void;
}

// Operations that can be performed on a customer.
export interface Inbound {
  // Creates a new customer.
  createCustomer(customer: Customer): Promise<Customer>;
  // Retrieve a customer by id.
  getCustomer(id: number): Promise<Customer>;
  // Return a page of customers using optional search filters.
  listCustomers(query: CustomerQuery): Promise<CustomerPage>;
}

// Stateful operations that can be performed on a customer.
export interface CustomerActor {
  // Creates the customer state.
  createCustomer(ctx: Context, customer: Customer): Promise<Customer>;
  // Retrieve the customer state.
  getCustomer(ctx: Context): Promise<Customer>;
}

export interface Outbound {
  // Saves a customer to the backend database
  saveCustomer(customer: Customer): Promise<void>;
  // Fetches a customer from the backend database
  fetchCustomer(id: number): Promise<Customer>;
  // Sends a customer creation event
  customerCreated(customer: Customer): Promise<void>;
  // Get customers from the database
  getCustomers(): IPublisher<Customer>;
  // Transform customers
  transformCustomers(
    prefix: string,
    source: Source<Customer>
  ): IPublisher<Customer>;
}

// Customer information.
export class Customer {
  // The customer identifer
  @Expose() id: number;
  // The customer's first name
  @Expose() firstName: string;
  // The customer's middle name
  @Expose() middleName: string | undefined;
  // The customer's last name
  @Expose() lastName: string;
  // The customer's email address
  @Expose() email: string;
  // The customer's address
  @Expose() address: Address;

  constructor({
    id = 0,
    firstName = "",
    middleName = null,
    lastName = "",
    email = "",
    address = new Address(),
  }: {
    id?: number;
    firstName?: string;
    middleName?: string | undefined;
    lastName?: string;
    email?: string;
    address?: Address;
  } = {}) {
    this.id = id;
    this.firstName = firstName;
    this.middleName = middleName;
    this.lastName = lastName;
    this.email = email;
    this.address = address;
  }
}

export class CustomerQuery {
  // The customer identifer
  @Expose() id: number | undefined;
  // The customer's first name
  @Expose() firstName: string | undefined;
  // The customer's middle name
  @Expose() middleName: string | undefined;
  // The customer's last name
  @Expose() lastName: string | undefined;
  // The customer's email address
  @Expose() email: string | undefined;
  @Expose() offset: number;
  @Expose() limit: number;

  constructor({
    id = null,
    firstName = null,
    middleName = null,
    lastName = null,
    email = null,
    offset = 0,
    limit = 100,
  }: {
    id?: number | undefined;
    firstName?: string | undefined;
    middleName?: string | undefined;
    lastName?: string | undefined;
    email?: string | undefined;
    offset?: number;
    limit?: number;
  } = {}) {
    this.id = id;
    this.firstName = firstName;
    this.middleName = middleName;
    this.lastName = lastName;
    this.email = email;
    this.offset = offset;
    this.limit = limit;
  }
}

export class CustomerPage {
  @Expose() offset: number;
  @Expose() limit: number;
  @Expose() items: Array<Customer>;

  constructor({
    offset = 0,
    limit = 0,
    items = new Array<Customer>(),
  }: { offset?: number; limit?: number; items?: Array<Customer> } = {}) {
    this.offset = offset;
    this.limit = limit;
    this.items = items;
  }
}

export class Nested {
  @Expose() foo: string;
  @Expose() bar: string;

  constructor({ foo = "", bar = "" }: { foo?: string; bar?: string } = {}) {
    this.foo = foo;
    this.bar = bar;
  }
}

// Address information.
export class Address {
  // The address line 1
  @Expose() line1: string;
  // The address line 2
  @Expose() line2: string | undefined;
  // The city
  @Expose() city: string;
  // The state
  @Expose() state: string;
  // The zipcode
  @Expose() zip: string;

  constructor({
    line1 = "",
    line2 = null,
    city = "",
    state = "",
    zip = "",
  }: {
    line1?: string;
    line2?: string | undefined;
    city?: string;
    state?: string;
    zip?: string;
  } = {}) {
    this.line1 = line1;
    this.line2 = line2;
    this.city = city;
    this.state = state;
    this.zip = zip;
  }
}

// Error response.
export class Error {
  // The detailed error message
  @Expose() message: string;

  constructor({ message = "" }: { message?: string } = {}) {
    this.message = message;
  }
}

File: example/customers/pkg/customers/interfaces.go
package customers

import (
	"context"
)

type ns struct{}

func (n *ns) Namespace() string {
	return "customers.v1"
}

func (n *ns) Version() string {
	return "0.1.0"
}

type LogicalAddress struct {
	Type string `json:"type,omitempty" msgpack:"type,omitempty"`
	ID   string `json:"id,omitempty" msgpack:"id,omitempty"`
}

func (a LogicalAddress) String() string {
	return a.Type + "/" + a.ID
}

type Context interface {
	context.Context
	Self() LogicalAddress
	Get(key string, dst interface{}) (bool, error)
	Set(key string, data interface{})
	Remove(key string)
}

// Operations that can be performed on a customer.
type Inbound struct {
	// Creates a new customer.
	CreateCustomer func(ctx context.Context, customer Customer) (*Customer, error)
	// Retrieve a customer by id.
	GetCustomer func(ctx context.Context, id uint64) (*Customer, error)
	// Return a page of customers using optional search filters.
	ListCustomers func(ctx context.Context, query CustomerQuery) (*CustomerPage, error)
}

// Stateful operations that can be performed on a customer.
type CustomerActor interface {
	// Creates the customer state.
	CreateCustomer(ctx Context, customer Customer) (*Customer, error)
	// Retrieve the customer state.
	GetCustomer(ctx Context) (*Customer, error)
}

type Outbound interface {
	SaveCustomer(ctx context.Context, customer Customer) error
	FetchCustomer(ctx context.Context, id uint64) (*Customer, error)
	CustomerCreated(ctx context.Context, customer Customer) error
}

// Customer information.
type Customer struct {
	ns
	// The customer identifer
	ID uint64 `json:"id" msgpack:"id"`
	// The customer's first name
	FirstName string `json:"firstName" msgpack:"firstName"`
	// The customer's middle name
	MiddleName *string `json:"middleName,omitempty" msgpack:"middleName,omitempty"`
	// The customer's last name
	LastName string `json:"lastName" msgpack:"lastName"`
	// The customer's email address
	Email string `json:"email" msgpack:"email"`
	// The customer's address
	Address Address `json:"address" msgpack:"address"`
}

func (c *Customer) Type() string {
	return "Customer"
}

type CustomerQuery struct {
	ns
	// The customer identifer
	ID *uint64 `json:"id,omitempty" msgpack:"id,omitempty"`
	// The customer's first name
	FirstName *string `json:"firstName,omitempty" msgpack:"firstName,omitempty"`
	// The customer's middle name
	MiddleName *string `json:"middleName,omitempty" msgpack:"middleName,omitempty"`
	// The customer's last name
	LastName *string `json:"lastName,omitempty" msgpack:"lastName,omitempty"`
	// The customer's email address
	Email  *string `json:"email,omitempty" msgpack:"email,omitempty"`
	Offset uint64  `json:"offset" msgpack:"offset"`
	Limit  uint64  `json:"limit" msgpack:"limit"`
}

func (c *CustomerQuery) Type() string {
	return "CustomerQuery"
}

type CustomerPage struct {
	ns
	Offset uint64     `json:"offset" msgpack:"offset"`
	Limit  uint64     `json:"limit" msgpack:"limit"`
	Items  []Customer `json:"items" msgpack:"items"`
}

func (c *CustomerPage) Type() string {
	return "CustomerPage"
}

type Nested struct {
	ns
	Foo string `json:"foo" msgpack:"foo"`
	Bar string `json:"bar" msgpack:"bar"`
}

func (n *Nested) Type() string {
	return "Nested"
}

// Address information.
type Address struct {
	ns
	// The address line 1
	Line1 string `json:"line1" msgpack:"line1"`
	// The address line 2
	Line2 *string `json:"line2,omitempty" msgpack:"line2,omitempty"`
	// The city
	City string `json:"city" msgpack:"city"`
	// The state
	State string `json:"state" msgpack:"state"`
	// The zipcode
	Zip string `json:"zip" msgpack:"zip"`
}

func (a *Address) Type() string {
	return "Address"
}

// Error response.
type Error struct {
	ns
	// The detailed error message
	Message string `json:"message" msgpack:"message"`
}

func (e *Error) Type() string {
	return "Error"
}
